crates/synthesis/src/discrete_basis/solovay_kitaev.rs

```diff
@@ -0,0 +1,383 @@
+//
+// Any modifications or derivative works of this code must retain this
+// copyright notice, and modified files need to carry a notice indicating
+// that they have been altered from the originals.
+
+use nalgebra::Matrix2;
+use numpy::{Complex64, PyReadonlyArray2};
+use pyo3::exceptions::{PyRuntimeError, PyValueError};
+use pyo3::types::PyString;
+use pyo3::{prelude::*, types::PyList};
+use qiskit_circuit::circuit_data::CircuitData;
+use qiskit_circuit::circuit_instruction::OperationFromPython;
+use qiskit_circuit::operations::{Operation, OperationRef, Param, StandardGate};
+
+use crate::discrete_basis::basic_approximations::{BasicApproximations, GateSequence};
+
+use super::basic_approximations::DiscreteBasisError;
+use super::math::{self, group_commutator_decomposition};
+
+/// A stateful implementation of Solovay Kitaev.
+///
+/// This generates the basic approximation set once as R-tree and re-uses it for
+/// each queried decomposition.
+#[pyclass]
+pub struct SolovayKitaevSynthesis {
+    /// The set of basic approximations.
+    basic_approximations: BasicApproximations,
+    /// Whether to perform runtime checks on the handled matrices/data.
+    do_checks: bool,
+}
+
+impl SolovayKitaevSynthesis {
+    /// Initialize a new instance by generating a set of basic approximations, given a list
+    /// of discrete standard gates and a depth. Returns an error if any gate is not discrete.
+    pub fn new(
+        basis_gates: &[StandardGate],
+        depth: usize,
+        tol: Option<f64>,
+        do_checks: bool,
+    ) -> Result<Self, DiscreteBasisError> {
+        let basic_approximations = BasicApproximations::generate_from(basis_gates, depth, tol)?;
+        Ok(Self {
+            basic_approximations,
+            do_checks,
+        })
+    }
+
+    /// Run the Solovay Kitaev algorithm, given the initial unitary as U(2) matrix.
+    ///
+    /// This matrix is given using [Complex64] numbers, which can limit the precision of the
+    /// algorithm. It is preferred to run the algorithm using [synthesize_std], which let's Qiskit
+    /// attempt to construct the underlying SO(3) matrix representation using [f64] at
+    /// quadruple precision, instead of inferring it from the double precision U(2) matrix.
+    pub fn synthesize_matrix(
+        &self,
+        matrix_u2: &Matrix2<Complex64>,
+        recursion_degree: usize,
+    ) -> Result<CircuitData, DiscreteBasisError> {
+        let input_sequence = GateSequence::from_u2(matrix_u2, self.do_checks);
+        self.synthesize_sequence(&input_sequence, recursion_degree)
+    }
+
+    /// Run the Solovay Kitaev algorithm on a standard gate + gate parameters.
+    ///
+    /// This attempts to directly construct the SO(3) matrix representation to minimize roundoff
+    /// errors. If unsuccessful, this falls back onto constructing the standard U(2) matrix
+    /// representation and converting it to SO(3).
+    pub fn synthesize_std(
```

Comment:
Is this comment still true without the support of `BigFloat`? Otherwise, we might not this additional flow.

Reply:
It is 🙂 In my benchmarks constructing the SO(3) representation directly reduces the error by a factor 2-4x.

Reply:
I'll resolve this as it is the same conversation as https://github.com/Qiskit/qiskit/pull/14203#discussion_r2106106385